core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java

```diff
@@ -400,41 +382,127 @@ public boolean isEmpty() {
+            .exceptionally(throwable -> {
+               log.debugf(throwable, "[IRAC] Failed to load entry to send to remote sites. It will be retried. State=%s", state);
+               state.retry();
+               return null;
+            }));
+   }
+
+   private CompletableSource sendUpdateBatch(Collection<? extends IracStateData> batch) {
+      int size = batch.size();
+      boolean trace = log.isTraceEnabled();
+
+      if (trace) {
+         log.tracef("[IRAC] Batch ready to send remote site with %d keys", size);
+      }
+
+      if (size == 0) {
+         if (trace) {
+            log.trace("[IRAC] Batch not sent, reason: batch is empty");
+         }
+         return Completable.complete();
+      }
+      IracPutManyCommand cmd = commandsFactory.buildIracPutManyCommand(size);
+      Collection<IracManagerKeyState> invalidState = new ArrayList<>(size);
+      Collection<IracManagerKeyState> validState = new ArrayList<>(size);
+      for (IracStateData data : batch) {
+         if (data.entry == null && data.tombstone == null) {
+            // there a concurrency issue where the entry and the tombstone do not exist (remove following by a put)
+            // the put will create a new state and the key will be sent to the remote sites
+            invalidState.add(data.state);
+            continue;
+         }
+         validState.add(data.state);
+         if (data.state.isExpiration()) {
+            cmd.addExpire(data.state.getKey(), data.tombstone);
+         } else if (data.entry == null) {
+            cmd.addRemove(data.state.getKey(), data.tombstone);
+         } else {
+            cmd.addUpdate(data.state.getKey(), data.entry.getValue(), data.entry.getMetadata(), data.entry.getInternalMetadata().iracMetadata());
+         }
+      }
+      IracResponseCollector rspCollector = new IracResponseCollector(commandsFactory.getCacheName(), validState, this::onBatchResponse);
+      try {
+         for (IracXSiteBackup backup : asyncBackups) {
+            if (takeOfflineManager.getSiteState(backup.getSiteName()) == SiteState.OFFLINE) {
+               continue; // backup is offline
+            }
+            rspCollector.dependsOn(backup, sendToRemoteSite(backup, cmd));
+         }
+      } catch (Throwable throwable) {
+         // marshalling error? JGroups error?
+         for (IracStateData data : batch) {
+            data.state.retry();
+         }
+         onUnexpectedThrowable(throwable);
+         return Completable.complete();
+      }
+
+      if (invalidState.isEmpty()) {
+         return Completable.fromCompletionStage(rspCollector.freeze());
+      }
+
+      if (trace) {
+         log.tracef("[IRAC] Removing %d invalid state(s)", invalidState.size());
+      }
+      invalidState.forEach(IracManagerKeyState::discard);
+      removeStateFromCluster(invalidState);
+      return Completable.fromCompletionStage(rspCollector.freeze());
```

Comment:
```suggestion
      if (!invalidState.isEmpty()) {
         if (trace) {
            log.tracef("[IRAC] Removing %d invalid state(s)", invalidState.size());
         }
         invalidState.forEach(IracManagerKeyState::discard);
         removeStateFromCluster(invalidState);
      }

      return Completable.fromCompletionStage(rspCollector.freeze());
```
Just a bit more compact.